### PR TITLE
Update OpenTelemetry SDK to the latest version (1.7.1)

### PR DESCRIPTION
**Pull request by joaopgrassi**
This PR updates the OTel SDK version to the latest. This update introduced several changes, most notably the change to use attributes instead of labels. The attributes now have types, but for now we'll only handle string values. All other values will be ignored with a warning log message.

- [x] Update to the latest SDK version
- [x] Update sample apps
- [x] Update README, and add a section informing users about the limitation with string-only attributes
- [x] Add a test using Views to ensure our export works with it

### CODE REVIEW

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/Serializer.java [hunk 1]
```diff
@@ -55,14 +63,23 @@
...
-    labels.forEach((k, v) -> dimensions.add(Dimension.create(k, v)));
+  static List<Dimension> toListOfDimensions(Attributes attributes) {
+    ArrayList<Dimension> dimensions = new ArrayList<>(attributes.size());
+
+    attributes.forEach(
+        (k, v) -> {
+          if (k.getType() == AttributeType.STRING) {
+            dimensions.add(Dimension.create(k.getKey(), String.valueOf(v)));
+          } else {
+            logger.warning(String.format(TEMPLATE_MSG_NON_SUPPORTED_ATTRIBUTE_TYPE, k.getType()));
```
Comment (pichlermarc) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/Serializer.java:74]: This will always invoke `String.format`, regardless of log-level. I have seen this in a few other places as well which were not part of your PR. 

I would suggest using [this overload of warning](https://docs.oracle.com/javase/8/docs/api/java/util/logging/Logger.html#warning-java.util.function.Supplier-) which takes a supplier, or [this overload of `log`](https://docs.oracle.com/javase/8/docs/api/java/util/logging/Logger.html#log-java.util.logging.Level-java.lang.String-java.lang.Object-).

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporter.java [hunk 2]
```diff
@@ -66,7 +66,7 @@ private DynatraceMetricExporter(
       URL url,
       String apiToken,
       String prefix,
-      Labels defaultDimensions,
+      Attributes defaultDimensions,
       Boolean enrichWithOneAgentMetaData) {
```
Comment (pichlermarc) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporter.java:70]: Not part of this PR, but is there a reason this is a boxed Boolean? Could throw NullPointerException when used in expressions when it is `null`.

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/Serializer.java (hunk 1 above)
Comment (joaopgrassi) [reply dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/Serializer.java]: Cool! I changed to use the "lazy" overload everywhere.

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporter.java (hunk 2 above)
Comment (joaopgrassi) [reply dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporter.java:70]: No idea tbh. Maybe @pirgeo knows?

Comment (pirgeo) [reply dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporter.java:70]: TBH, I am not sure. Probably just an oversight on my part, you can try and replace it and see if something breaks 😇

Comment (joaopgrassi) [reply dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporter.java]: I changed to use the value type one. Everything still works.

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java [hunk 3]
```diff
@@ -167,10 +167,10 @@ public void addDefaultDimensions() throws IOException {
   }
 
   @Test
-  public void testWithLabels() throws IOException {
+  public void testWithAttributes() throws IOException {
```
Comment (pichlermarc) [inline dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java:171]: Nit: some of the test have the `public` modifier while others don't. It is not necessary to have the `public` modifier starting from JUnit 5. Could remove it for a more uniform look. :)

File: dynatrace/build.gradle [hunk 4]
```diff
@@ -20,14 +20,14 @@ plugins {
 }
 
 group = "com.dynatrace.opentelemetry.metric"
-version = "0.3.0-alpha"
+version = "0.4.0-alpha"
 
 repositories {
     mavenCentral()
     jcenter()
```
Comment (pichlermarc) [inline dynatrace/build.gradle:27]: JCenter will stop serving packages after Febuary 1st, 2022. Should we remove it now?

File: example/src/main/java/com/dynatrace/opentelemetry/metric/example/DynatraceExporterExample.java [hunk 5]
```diff
@@ -56,46 +58,40 @@ public static void main(String[] args) throws Exception {
...
 
-    // Create a metrics producer and a meter provider (SdkMeterProvider is both). As noted in the
-    // documentation (https://opentelemetry.io/docs/java/manual_instrumentation/, under the metrics
-    // section), the APIs for acquiring a MeterProvider are in flux and the example below is
-    // probably outdated. Please consult the OpenTelemetry documentation for more information on
+    // Creates the meter provider, configuring the metric reader and our exporter in it. As noted in
+    // the documentation (https://opentelemetry.io/docs/java/manual_instrumentation/, under the
+    // metrics section), the APIs for acquiring a MeterProvider are in flux, so the example below
+    // might be outdated. Please consult the OpenTelemetry documentation for more information on
     // the preferred way to acquire a MeterProvider.
```
Comment (pirgeo) [inline example/src/main/java/com/dynatrace/opentelemetry/metric/example/DynatraceExporterExample.java:65]:
```suggestion
    // Creates the meter provider, configuring the metric reader and the Dynatrace exporter.
```

File: example/src/main/java/com/dynatrace/opentelemetry/metric/example/DynatraceExporterExample.java [hunk 6]
```diff
@@ -56,46 +58,40 @@ public static void main(String[] args) throws Exception {
...
+    SdkMeterProvider meterProvider =
+        SdkMeterProvider.builder()
+            .registerMetricReader(PeriodicMetricReader.create(exporter, Duration.ofMillis(60000)))
+            .buildAndRegisterGlobal();
+
+    // Get or create a named meter instance. If a reference to the MeterProvider ist kept,
+    // meterProvider.get(...) would do the same.
     Meter meter =
-        GlobalMeterProvider.getMeter(DynatraceExporterExample.class.getName(), "0.1.0-beta");
+        GlobalMeterProvider.get().get(DynatraceExporterExample.class.getName(), "0.1.0-beta", "");
```
Comment (pirgeo) [inline example/src/main/java/com/dynatrace/opentelemetry/metric/example/DynatraceExporterExample.java:74]:
```suggestion
    Meter meter =
        GlobalMeterProvider.get()
            .meterBuilder("com.example.my_instrumentation_library")
            .setInstrumentationVersion("0.1.0-alpha")
            .build();
```

File: example/src/main/java/com/dynatrace/opentelemetry/metric/example/DynatraceExporterExample.java [hunk 7]
```diff
@@ -56,46 +58,40 @@ public static void main(String[] args) throws Exception {
...
     LongCounter counter =
         meter
-            .longCounterBuilder("example_counter")
+            .counterBuilder("example_counter")
             .setDescription("Just some counter used as an example")
             .setUnit("1")
             .build();
 
-    // Use a bound counter with a pre-defined label set
+    // Create a bound counter with a pre-defined attributes in it
```
Comment (pirgeo) [inline example/src/main/java/com/dynatrace/opentelemetry/metric/example/DynatraceExporterExample.java:84]:
```suggestion
    // Create a bound counter with a pre-defined attribute
```

File: example/src/main/java/com/dynatrace/opentelemetry/metric/example/DynatraceExporterExample.java [hunk 8]
```diff
@@ -56,46 +58,40 @@ public static void main(String[] args) throws Exception {
...
 
     while (true) {
-      // Record data with bound labels
+      // Record data with bound attributes
       someWorkCounter.add(random.nextInt(5));
 
-      // Or record data on unbound counter and explicitly specify the label set at call-time
-      counter.add(random.nextInt(10), Labels.of("environment", "testing"));
-      counter.add(random.nextInt(20), Labels.of("environment", "staging"));
+      // Or record data on unbound counter and explicitly specify the attributes at call-time
```
Comment (pirgeo) [inline example/src/main/java/com/dynatrace/opentelemetry/metric/example/DynatraceExporterExample.java:92]:
```suggestion
      // Or record data on an unbound counter and explicitly specify the attributes at call-time
```

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java [hunk 9]
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright 2020 Dynatrace LLC
+ * Copyright 2021 Dynatrace LLC
```
Comment (pirgeo) [inline dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java:2]: I think this should be the year when it was initially released, so it stays 2020.

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java [hunk 10]
```diff
@@ -136,9 +141,519 @@ public void testAddPrefix() throws IOException {
...
+            .build();
+
+    CompletableResultCode result = metricExporter.export(Collections.singleton(md), connection);
+
+    verify(connection).setRequestMethod("POST");
+    verify(connection).setRequestProperty("Authorization", "Api-Token mytoken");
+    verify(connection).setRequestProperty("Content-Type", "text/plain; charset=utf-8");
+
+    String a = bos.toString();
+    assertThat(a)
```
Comment (pirgeo) [inline dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java:195]:
```suggestion
    // even tough Attributes are sorted internally, they might be serialized in an unsorted manner. This is due to de-duplication.
    assertThat(a)
```

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java [hunk 11]
```diff
@@ -136,9 +141,519 @@ public void testAddPrefix() throws IOException {
...
+    Collection<DoublePointData> doublePointDataCollection =
+        new ArrayList<DoublePointData>() {
+          {
+            add(
+                DoublePointData.create(
+                    1619687639000000000L, 1619687659000000000L, Attributes.empty(), 100.3));
+            add(
+                DoublePointData.create(
+                    1619687639000000000L, 1619687659000000000L, Attributes.empty(), 300.6));
+            add(DoublePointData.create(0L, 0L, Attributes.empty(), 500.8));
```
Comment (pirgeo) [inline dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java:325]: Is there a reason why you didn't put timestamps on these?

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java [hunk 12]
```diff
@@ -136,9 +141,519 @@ public void testAddPrefix() throws IOException {
...
+            .setApiToken("mytoken")
+            .setUrl(connection.getURL())
+            .build();
+
+    CompletableResultCode result =
+        metricExporter.export(Collections.singleton(metricData), connection);
+
+    verify(connection).setRequestMethod("POST");
+    verify(connection).setRequestProperty("Authorization", "Api-Token mytoken");
+    verify(connection).setRequestProperty("Content-Type", "text/plain; charset=utf-8");
```
Comment (pirgeo) [inline dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java:401]:
```suggestion
    verifyConnectionProperties(connection);
```
i wonder if its worth putting these three lines into a separate function, they are repeated a lot, right?

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java [hunk 13]
```diff
@@ -150,26 +665,51 @@ public void addDefaultDimensions() throws IOException {
...
+    testClock.advance(Duration.ofSeconds(1));
+
+    counter.add(300);
+    testClock.advance(Duration.ofSeconds(1));
+    CompletableResultCode result = sdkMeterProvider.forceFlush();
+
+    assertTrue(result.isSuccess());
     assertEquals(
-        "name,default=value,dt.metrics.source=opentelemetry count,delta=194.0 1619687659000",
+        "testSum,dt.metrics.source=opentelemetry count,delta=200 1557212401000testSum,dt.metrics.source=opentelemetry count,delta=300 1557212403000",
```
Comment (pirgeo) [inline dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java:702]: Why are these concatenated with a comma, and not with a newline?

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java [hunk 14]
```diff
@@ -183,14 +723,52 @@ public void testWithLabels() throws IOException {
...
+    testClock.advance(Duration.ofSeconds(1));
+
+    counter.add(200);
+    testClock.advance(Duration.ofSeconds(1));
+    CompletableResultCode result = sdkMeterProvider.forceFlush();
+
+    assertTrue(result.isSuccess());
     assertEquals(
-        "name,dt.metrics.source=opentelemetry,label1=val1,label2=val2 count,delta=194.0 1619687659000",
+        "testSum,dt.metrics.source=opentelemetry count,delta=100 1557212400000testSum,dt.metrics.source=opentelemetry count,delta=200 1557212402000",
```
Comment (pirgeo) [inline dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java:754]: These are not separated at all

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/SerializerTest.java [hunk 15]
```diff
@@ -32,31 +33,62 @@
...
+    Attributes attributes =
+        Attributes.builder().put("l1", 1).put("l2", 1.5).put("l3", true).build();
+
     DimensionList expected = DimensionList.create();
-    DimensionList actual = Serializer.fromLabels(labels);
+
+    DimensionList actual = Serializer.fromAttributes(attributes);
 
     assertEquals(expected.getDimensions(), actual.getDimensions());
     assertNotSame(expected, actual);
```
Comment (pirgeo) [inline dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/SerializerTest.java:60]: Could you also assert empty, i think it would make it more clear.

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/SerializerTest.java [hunk 16]
```diff
@@ -32,31 +33,62 @@
...
   }
 
   @Test
-  void fromLabelsNormalize() {
-    Labels labels = Labels.of("~~!123", "test", "!!test2", "test2");
-    DimensionList actual = Serializer.fromLabels(labels);
+  void fromTypedAttributesArrayValues() {
+    Attributes attributes = Attributes.builder().put("l1", "l1v1", "l1v2").put("l2", 1, 2).build();
```
Comment (pirgeo) [inline dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/SerializerTest.java:65]: Could also add one for booleans or doubles if you want, but i think clear from the context anyway.

**Review by pirgeo — CHANGES_REQUESTED** [review]

File: README.md [hunk 17]
```diff
@@ -50,52 +55,99 @@ DynatraceMetricExporter exporter =
...
+// Then, we create the MeterProvider, making sure to register our PeriodicMetricReader:
+SdkMeterProvider meterProvider = 
+    SdkMeterProvider.builder()
+    .registerMetricReader(metricReader)
+    .buildAndRegisterGlobal();
+
+// OR you can also use a shorter version:
+SdkMeterProvider meterProvider =
+    SdkMeterProvider.builder()
+    .registerMetricReader(PeriodicMetricReader.create(exporter, Duration.ofMillis(60000)))
```
[the remark this answers is not in the document]
Comment (joaopgrassi) [reply README.md:74]: It's the same, I believe just a shorter way. If you don't need the factory then you can use the shorter one. I think maybe it's better to use the shorter than the first as it's less code?

File: README.md [hunk 18]
```diff
@@ -108,25 +160,31 @@ The Dynatrace API token to be used by the exporter is specified using `setApiTok
...
 #### Default Dimensions
 
-The `setDefaultDimensions` method can be used to optionally specify a `Labels` object, which will be added as additional dimensions to all data points.
-`Labels` represent key-value pairs.
-Dimension keys will be normalized, de-duplicated, and only one dimension value per key will be sent to the server.
+The `setDefaultDimensions` method can be used to optionally specify a
+[Attributes](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/common/Attributes.java) object,
+which will be added as additional dimensions to all data points. The `Attributes` interface represents key-value pairs.
+
+Dimension keys will be normalized, prior to be sent to the server.
```
[the remark this answers is not in the document]
Comment (joaopgrassi) [reply README.md:184]: They are already de-duped by the SDK when using the Attributes built-in impl. So we don't need to mention this?

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/InMemoryDtMetricReader.java [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.Collection;
+import java.util.Collections;
+
+/**
+ * Inspired on
+ * https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/testing/InMemoryMetricReader.java
+ * A reader that calls the exporter whenever it receives a flush call from the MeterProvider. Useful
+ * for testing the exporter working with a configured MeterProvider
+ */
+class InMemoryDtMetricReader implements MetricReader, MetricReaderFactory {
```
[the remark this answers is not in the document]
Comment (joaopgrassi) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/InMemoryDtMetricReader.java:18]: The InMemory from OTel doesn't call our exporter and simply expose a way to collect the metrics. Since I wanted to test our exporter with the MeterProvider this was a way I find out.

I agree with the DT, will change it to something else :)

File: README.md [hunk 20]
```diff
@@ -10,7 +10,8 @@
 The general setup of OpenTelemetry Java is explained in the official [Getting Started Guide](https://opentelemetry.io/docs/java/manual_instrumentation/).
 Using the Metrics API is explained in the [Metrics section](https://opentelemetry.io/docs/java/manual_instrumentation/#metrics-alpha-only).
 
-To include the Dynatrace OpenTelemetry Metrics exporter in a Gradle build, for example, use the following in your `settings.gradle` and `build.gradle`:
+To include the Dynatrace OpenTelemetry Metrics exporter in a Gradle build, for example,
```
[the remark this answers is not in the document]
Comment (joaopgrassi) [reply README.md:13]: Yeah I ran markdown lint and I started fixing some long lines but decided to stop midway as it was adding too many changes that would make the review harder.

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java (hunk 3 above)
Comment (joaopgrassi) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java]: Done. :)

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/SerializerTest.java [hunk 21]
```diff
@@ -598,21 +630,24 @@ void createInvalidDoubleHistogramLines() {
...
                 10.234,
                 Arrays.asList(Double.NEGATIVE_INFINITY, 1.2d, 3.4d, 5.6d),
                 Arrays.asList(0L, 2L, 1L, 3L, 0L)));
   }
 
   @Test
   void TestGetMinFromBoundaries() {
+
+    Attributes attributes =
+        Attributes.of(stringKey("attr1"), "value1", stringKey("attr2"), "value2");
```
[the remark this answers is not in the document]
Comment (joaopgrassi) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/SerializerTest.java:643]: Nope. Removed.

File: README.md [hunk 22]
```diff
@@ -50,52 +55,99 @@ DynatraceMetricExporter exporter =
...
 
-The interval in which metrics are exported can be set on the `IntervalMetricReader` (see above).
+The interval in which metrics are exported can be set on the `PeriodicMetricReader` (see above).
 In the example case above, metrics are exported every 60 seconds.
 
 Once metrics are reported using the Metrics API, data will be exported to Dynatrace:
 
 ```java
-Meter meter = GlobalMeterProvider.getMeter("com.example.my_instrumentation_library", "0.1.0-alpha");
+Meter meter = GlobalMeterProvider.get().get("com.example.my_instrumentation_library", "0.1.0-alpha", "");
```
[the remark this answers is not in the document]
Comment (joaopgrassi) [reply README.md:84]: Yeah I was also confused. I will this version, more clear :)

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/Serializer.java [hunk 23]
```diff
@@ -55,14 +63,24 @@
     return builder;
   }
 
-  static List<Dimension> toListOfDimensions(Labels labels) {
-    ArrayList<Dimension> dimensions = new ArrayList<>(labels.size());
-    labels.forEach((k, v) -> dimensions.add(Dimension.create(k, v)));
+  static List<Dimension> toListOfDimensions(Attributes attributes) {
+    ArrayList<Dimension> dimensions = new ArrayList<>(attributes.size());
```
[the remark this answers is not in the document]
Comment (joaopgrassi) [reply dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/Serializer.java:67]: Good point. The default size of `new ArrayList` seems to be 10. If it's common to add more than 10 than using the default might make more sense as if they are dropped we don't create a bigger array than needed. But if it's common to add, say 5 attributes than even if we drop some it's still "more efficient" to allocate?

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java (hunk 13 above)
Comment (joaopgrassi) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java]: It was because the same instance of `ByteArrayOutputStream` was used across the flushes. I refactored it now so it is reset after each call to `flush`. Now the lines are asserted individually.

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java (hunk 14 above)
Comment (joaopgrassi) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java]: Same as above. Now each line is asserted individually.

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/SerializerTest.java (hunk 16 above)
Comment (joaopgrassi) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/SerializerTest.java:64]: Done.

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/SerializerTest.java (hunk 15 above)
Comment (joaopgrassi) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/SerializerTest.java:59]: Done. Changed in another test that did the same. More clear for sure.

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java (hunk 11 above)
Comment (joaopgrassi) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java:327]: Not really, I basically copy the lines from the `createDoubleSumLinesCumulative` in the Serializer tests.

File: README.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (pirgeo) [reply README.md:184]: I'd still mention it, since we combine default dimensions, static dimensions and the dimensions from the attributes. This might lead to people wondering why attributes were dropped, and having this line here makes it clear IMO

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/Serializer.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (pirgeo) [reply dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/Serializer.java:67]: I'd say 5 attributes is a good guess, but i might be totally wrong. I guess well leave it as it is then.

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java (hunk 11 above)
Comment (pirgeo) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java:327]: It leads to different lines, one with and one without timestamp, right? So i guess we can leave it like this, as this will then also test serialization with and without timestamp.

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java (hunk 12 above)
Comment (joaopgrassi) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java]: Yes. Done :)

File: README.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (joaopgrassi) [reply README.md]: OK I fixed all markdownlint warnings in the README.

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java (hunk 11 above)
Comment (joaopgrassi) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporterTest.java:317]: Exactly. I'll add a comment so it's clear it was intentional.

File: README.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (joaopgrassi) [reply README.md:85]: I'll keep the shorter version for brevity. Rephrased the comment a bit more.

File: dynatrace/build.gradle (hunk 4 above)
Comment (joaopgrassi) [reply dynatrace/build.gradle:27]: I'm not really familiar with this, but sounds like a good idea then?

**Review by pirgeo — APPROVED** [review]
Looking good IMO!

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/CumulativeToDeltaConverter.java [hunk 24]
```diff
@@ -137,18 +133,33 @@ void reset() {
...
-  private static String getSortedLabelsString(Labels labels) {
-    if (labels.isEmpty()) {
+  private static String createIdentifier(String name, Attributes attributes, String type) {
+    // (\u001d = ASCII group separator)
+    return String.format("%s\u001d%s\u001d%s", name, getSortedAttributesString(attributes), type);
+  }
+
+  /**
+   * The implementation of the {@link Attributes} interface shipped with OpenTelemetry ({@link
+   * ArrayBackedAttributes}) guarantees that the elements are sorted by their key. If a different
```
[the remark this answers is not in the document]
Comment (joaopgrassi) [reply dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/CumulativeToDeltaConverter.java:143]: From my understanding of the SDK, yes. It create an instance of `ArrayBackedAttributes` https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/common/ArrayBackedAttributes.java#L26, which then calls the base ctor that in turn does the mergeSort and dedup:     return dedupe(data, keyComparator);

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/CumulativeToDeltaConverterTest.java [hunk 25]
```diff
@@ -1,8 +1,22 @@
+/**
+ * Copyright 2020 Dynatrace LLC
```
[the remark this answers is not in the document]
Comment (joaopgrassi) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/CumulativeToDeltaConverterTest.java:2]: Ah ok I got confused before I thought you meant it should be in the year it was release.

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/MockedMetricReader.java [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.Collection;
+import java.util.Collections;
+
+/**
+ * Inspired by
+ * https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/testing/InMemoryMetricReader.java
+ * A reader that calls the exporter whenever it receives a flush call from the MeterProvider. Useful
+ * for testing the exporter working with a configured MeterProvider
+ */
+class MockedMetricReader implements MetricReader, MetricReaderFactory {
```
[the remark this answers is not in the document]
Comment (joaopgrassi) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/MockedMetricReader.java:18]: Renamed to `TestMetricReader`

File: dynatrace/build.gradle (hunk 4 above)
Comment (joaopgrassi) [reply dynatrace/build.gradle]: Ok, removed.

**Review by pichlermarc — APPROVED** [review]
Looking good :)

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/CumulativeToDeltaConverterTest.java [hunk 27]
```diff
@@ -212,4 +212,30 @@ void testDoubleAndLongDoNotInterfere() {
     assertThat(converter.convertDoubleTotalToDelta("test", createDoublePointData(200.5)))
         .isCloseTo(100.3, offset);
   }
+
+  @Test
+  void testAttributesAreDeDuplicated() {
+    // a delta is correctly calculated for the second metric, meaning that these two metrics are
+    // considered equal by the delta calculation. The duplicated attribute `attr1 is removed.
```
Comment (pirgeo) [inline dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/CumulativeToDeltaConverterTest.java:219]:
```suggestion
    // considered equal by the delta calculation. The duplicated attribute attr1 is removed.
```

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/CumulativeToDeltaConverterTest.java [hunk 28]
```diff
@@ -212,4 +212,30 @@ void testDoubleAndLongDoNotInterfere() {
...
+    assertThat(
+            converter.convertLongTotalToDelta(
+                "test",
+                createLongPointData(
+                    100L, 0, Attributes.of(stringKey("attr1"), "v1", stringKey("attr2"), "v2"))))
+        .isNull();
+    assertThat(
+            converter.convertLongTotalToDelta(
+                "test",
+                createLongPointData( // has 'attr1' twice
```
Comment (pirgeo) [inline dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/CumulativeToDeltaConverterTest.java:229]: If we use two different attribute values we could not assert here since the behavior is then undefined and we have no idea what would happen, right?

Comment (joaopgrassi) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/CumulativeToDeltaConverterTest.java:229]: We do have an idea what happens. The `attributes` from OTel does the de-duplication and the last values wins. So if we have `attr1=v1` and `attr1=v2`, the final attributes list will contain `attr1=v2`. 

This will "break" the metric identification if it was reported first without `v2` and later reported with the duplicated attribute. Our code will think it's a new metric.

**Review by pirgeo — APPROVED** [review]

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/CumulativeToDeltaConverterTest.java (hunk 28 above)
Comment (pirgeo) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/CumulativeToDeltaConverterTest.java:229]: Okay, so that means we could use `attr1=something random` and then `attr1=v1` here in the second metric point in order to make sure that the attributes are deduplicated based on their keys alone, right?

Comment (joaopgrassi) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/CumulativeToDeltaConverterTest.java]: Yes, correct. I added it now + another test to only make sure the OTel SDK does the correct sorting and deduplication.
